frontend/lib/api/ApiClient.ts

```diff
@@ -1,139 +1,141 @@
-          ApiResponseStatus.ServerError,
-          response.status,
-          `Server response parse error: ${response.status}`,
-        );
+      if (response.status >= 500 && response.status <= 599 && isError) {
+        return new ApiError(ApiResponseStatus.ServerError, response.status, body.error);
       }
+
+      return new ApiError(
+        ApiResponseStatus.ServerError,
+        response.status,
+        `Unexpected response status: ${response.status}`,
+      );
+    } catch (e) {
+      console.error("Error parsing response", e);
+
+      return new NetworkError((e as Error).message || "Network error");
     }
+  }
 
+  // handle a response without a body, and return an error when there is a non-2xx status or a non-empty body
+  async handleEmptyBody<T>(response: Response): Promise<ApiResult<T>> {
     const body = await response.text();
 
+    if (response.status >= 400 && response.status <= 499) {
+      return new ApiError(ApiResponseStatus.ClientError, response.status, body);
+    }
+
+    if (response.status >= 500 && response.status <= 599) {
+      return new ApiError(ApiResponseStatus.ServerError, response.status, body);
+    }
+
     if (body.length > 0) {
-      const message = `Unexpected data from server: ${body}`;
       console.error("Unexpected data from server:", body);
+
+      const message = `Unexpected data from server: ${body}`;
       return new ApiError(ApiResponseStatus.ServerError, response.status, message);
     }
 
-    if (response.status >= 200 && response.status <= 299) {
+    if (response.ok) {
       return {
         status: ApiResponseStatus.Success,
         code: response.status,
         data: body as T,
       };
     }
 
-    if (response.status >= 400 && response.status <= 499) {
-      return new ApiError(ApiResponseStatus.ClientError, response.status, body);
-    }
-
-    if (response.status >= 500 && response.status <= 599) {
-      return new ApiError(ApiResponseStatus.ServerError, response.status, body);
-    }
-
     return new ApiError(
       ApiResponseStatus.ServerError,
       response.status,
       `Unexpected response status: ${response.status}`,
     );
   }
 
-  async postRequest<T>(path: string, requestBody?: object): Promise<ApiResult<T>> {
-    let requestInit: RequestInit = {
-      method: "POST",
-    };
-    if (requestBody) {
-      requestInit = {
-        ...requestInit,
+  // perform a HTTP request and handle the response
+  async request<T>(
+    method: RequestMethod,
+    path: string,
+    abort?: AbortController,
+    requestBody?: object,
+  ): Promise<ApiResult<T>> {
+    try {
+      const response = await fetch(path, {
+        method,
         headers: {
-          "Content-Type": "application/json",
+          Accept: MIME_JSON,
         },
-        body: JSON.stringify(requestBody),
-      };
-    }
-    const response = await fetch(path, requestInit);
+        signal: abort?.signal,
+        ...this.setRequestBodyAndHeaders(requestBody),
```

Comment:
This will override the headers above

Reply:
That's also the intention. If we pass a body the content-type header should also be added.

Reply:
Ah I understand now, thanks